mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,616 @@
+                let mtt_l1_page =
+                    &mut mtt_core_locked.mtt_l1_page_pool_base[l1_pool_index].base[0..];
+                let value = mtt_l1_page[l1_index];
+                let bit_value = if memory_type == MttMemoryType::NonConfidential {
+                    0u64
+                } else {
+                    1u64
+                };
+                // Clear the original 2-bits and OR in the new value
+                let value = (value & !(3 << sub_index)) | (bit_value << sub_index);
+                mtt_l1_page[l1_index] = value;
+                Ok(())
+            }
+            _ => Err(Error::Non4KL2Entry),
+        }
+    }
+
+    // Returns the memory type for physical address.
+    // The page can be mapped at any granularity in the MTT.
+    fn get_memory_type(&self) -> Result<MttMemoryType> {
+        match self.l2_entry {
+            MttNonConfidential1G => Ok(MttMemoryType::NonConfidential),
+            MttConfidential1G => Ok(MttMemoryType::Confidential),
+            Mtt2MBPages(value) => {
+                let addr_bits = self.get_phys_addr().bits();
+                let aligned_2mb = PageSize::Size2M.round_down(addr_bits);
+                let aligned_64mb = addr_bits & !((1024 * 1024 * 64) - 1);
+                let bit_index = (aligned_2mb - aligned_64mb) / PageSize::Size2M as u64;
+                if (value & (1 << bit_index)) == MttMemoryType::Confidential as u64 {
+                    Ok(MttMemoryType::Confidential)
+                } else {
+                    Ok(MttMemoryType::NonConfidential)
+                }
+            }
+            MttL1Dir(_) => self.get_4k_page_memory_type(),
+        }
+    }
+
+    // Updates the memory type for the physical address in the MTT.
+    // The MTT update is skipped if the page is already contained in a spanning
+    // range of the same type.
+    // The desired page size is assumed to be 4K, and larger spans are broken
+    // down into smaller regions.
+    fn set_memory_type(&mut self, memory_type: MttMemoryType) -> Result<()> {
+        let is_confidential = matches!(self.get_memory_type()?, MttMemoryType::Confidential);
+        // If the address is already spanned by a range of the same memory type
+        // there's nothing to do. If the memory types differ, we may have to split
+        // the spanning range into smaller regions.
+        if (memory_type == MttMemoryType::Confidential && !is_confidential)
+            || (memory_type == MttMemoryType::NonConfidential && is_confidential)
+        {
+            match self.get_l2_entry_type() {
+                MttConfidential1G | MttNonConfidential1G => {
+                    self.convert_1gb_to_64mb_range()?;
+                    self.convert_64mb_to_l1_dir()?;
+                    self.set_4k_page_memory_type(memory_type)?;
+                }
+                Mtt2MBPages(_) => {
+                    self.convert_64mb_to_l1_dir()?;
+                    self.set_4k_page_memory_type(memory_type)?;
```

Comment:
@atulkharerivos question: where is the reverse handled - where a set of 4KB pages are reclaimed to non-confidential?

Reply:
@rsahita See `fn set_4k_page_memory_type(&mut self, memory_type: MttMemoryType)`  -- the `MemoryType` can be one of these:

```
pub enum MttMemoryType {
    /// A non-confidential memory range
    NonConfidential = 0,
    /// A confidential memory range
    Confidential = 1,
}
```

Note that this just updates the type in the MTT L2 and we'll have to have orchestration at a higher level for scrubbing etc. In practice, the VMM will likely create a pool of confidential pages, and will reuse them to avoid transitioning back and forth. @atishp04 mentioned that this an optimization that isn't currently in kvm, but can be added.